File: servo.cpp
/*servo.cpp
 * Author: Jesse Offei-Nkansah
 * Date: 4.8.2018
 * Implementation file for Servo class
 */

#include <iostream>
#include <pigpio.h>
#include "Servo.h"

Servo::Servo(int gpio_, int minPulse_, int maxPulse_, int pw_, int pwInc_){
	if (gpio_ >= 2 && gpio_ <= 27){
		gpio = gpio_;
	}else{
		gpio = -1;
	}
	
	if (minPulse_ >= 1000 && minPulse_ < 2000){
		minPulse = minPulse_;
	}else{
		minPulse = 1010;
	}
	
	if (maxPulse_ > minPulse && maxPulse_ <= 2000){
		maxPulse = maxPulse_;
	}else{
		maxPulse = 1800;
	}
	
	if(pw_ >= minPulse && pw_ <= maxPulse){
		pw = pw_;
	}else{
		pw = 1500;	
	}
	
	if(pwInc_ <= maxPulse - minPulse){
		pwInc = pwInc_;
	}else{
		pwInc = 15;
	}
}

Servo::Servo(int gpio_, int pw_, int pwInc_)
	: minPulse(1010), maxPulse(1800)
{
	if (gpio_ >= 2 && gpio_ <= 27){
		gpio = gpio_;
	}else{
		gpio = -1;
	}
	
	if(pw_ >= minPulse && pw_ <= maxPulse){
		pw = pw_;
	}else{
		pw = 1500;	
	}
	
	if(pwInc_ <= (maxPulse - minPulse)/2){
		pwInc = pwInc_;
	}else{
		pwInc = 15;
	}
}

Servo::Servo(int gpio_)
	: minPulse(1010), maxPulse(1800), pw(1500), pwInc(15)
{
	if (gpio_ >= 2 && gpio_ <= 27){
		gpio = gpio_;
	}else{
		gpio = -1;
	}
}


void Servo::setGPIO(int gpio_){
	if (gpio_ >= 2 && gpio_ <= 27){
		gpio = gpio_;
	}else{
		std::cerr << "Invalid GPIO.  No Change." << std::endl;
	}
}

int Servo::getGPIO(){
	return gpio;
}

void Servo::setMinPulse(int minPulse_){
	if (minPulse_ >= 1000 && minPulse_ < 2000){
		minPulse = minPulse_;
	}else{
		std::cerr << "Invalid Min Pulse.  No Change." << std::endl;
	}
}

int Servo::getMinPulse(){
	return minPulse;
}

void Servo::setMaxPulse(int maxPulse_){
	if (maxPulse_ > minPulse && maxPulse_ <= 2000){
		maxPulse = maxPulse_;
	}else{
		std::cerr << "Invallid Max Pulse. No Change." << std::endl;
	}
}

int Servo::getMaxPulse(){
	return maxPulse;
}

void Servo::setPulseWidth(int pw_){
	pw = pw_;
}

int Servo::getPulseWidth(){
	return pw;
}

void Servo::setIncrement(int pwInc_){
	pwInc = pwInc_;
}

int Servo::getIncrement(){
	return pwInc;
}

void Servo::pulse(){
	gpioServo(gpio, pw);
}

void Servo::increment(){
	pw += pwInc;
}

void Servo::decrement(){
	pw -= pwInc;
}

void Servo::servoLeft(){
	if(pwInc < 0){
		pw += (pwInc * -1);
	}else{
		pw += pwInc;
	}
}

void Servo::servoRight(){
	if(pwInc > 0){
		pw += (pwInc * -1);
	}else{
		pw += pwInc;
	}
}

Servo Servo::operator =(Servo& rhs){
	gpio = rhs.gpio;
	minPulse = rhs.minPulse;
	maxPulse = rhs.maxPulse;
	pw = rhs.pw;
	pwInc = rhs.pwInc;
}

bool Servo::operator ==(Servo& rhs){
	if(gpio == rhs.gpio && minPulse == rhs.minPulse &&
		maxPulse == rhs.maxPulse && pw == rhs.pw &&
		pwInc = rhs.pwInc){
		return true;
	}else{
		return false;
	}
}

bool Servo::operator !=(Servo& rhs){
	if(gpio == rhs.gpio && minPulse == rhs.minPulse &&
		maxPulse == rhs.maxPulse && pw == rhs.pw &&
		pwInc = rhs.pwInc){
		return false;
	}else{
		return true;
	}
}

File: PyoConnect_v2.0/servo.py
#servo.py
# Author: Jesse Offei-Nkansah
# Date: 4.10.2018
# Creates a servo class for controlling servos on GPIO

class Servo:
    """This class defines a servo object referenced by its GPIO location"""

    __gpio = None
    __pulse = None
    __maxPulse = None
    __minPulse = None
    __pwInc = None

    def __init__(self, gpio=2, pulse=1500, maxPulse=1850, minPulse=1080, pwInc=15):
        if (not self.set_gpio(gpio)):
            print("Setting to gpio 2")
            self.__gpio = 2

        if (not self.set_minPulse(minPulse)):
            print("Setting to 1080")
            self.__minPulse = 1080

        if (not self.set_maxPulse(maxPulse)):
            print("Setting to 1850")
            self.__maxPulse = 1850

        if(not self.set_pulse(pulse)):
            print("Setting to 1500")
            self.__pulse = 1500

        if(not self.set_pwInc(pwInc)):
            print("Setting to 15")
            self.__pwInc = 15

    def get_gpio(self):
        return self.__gpio

    def set_gpio(self, gpio):
        if (gpio >= 2) and (gpio <= 27):
            self.__gpio = gpio
            return True
        else:
            print("GPIO cannot be set to " + gpio + ".")
            return False

    def get_minPulse(self):
        return self.__minPulse

    def set_minPulse(self, minPulse):
        if (minPulse >= 1000) and (minPulse < 2000):
            self.__minPulse = minPulse
            return True
        else:
            print("Minimim Pulse cannot be set to " + minPulse + ".")
            return False

    def get_maxPulse(self):
        return self.__maxPulse

    def set_maxPulse(self, maxPulse):
        if (maxPulse > self.__minPulse) and (maxPulse <= 2000):
            self.__maxPulse = maxPulse
            return True
        else:
            print("Maximim Pulse cannot be set to " + maxPulse + ".")
            return False

    def get_pulse(self):
        return self.__pulse

    def set_pulse(self, pulse):
        if (pulse >= self.__minPulse) and (pulse <= self.__maxPulse):
            self.__pulse = pulse
            return True
        else:
            print("Pulse cannot be set to " + pulse + ".")
            return False

    def get_pwInc(self):
        return self.__pwInc

    def set_pwInc(self, pwInc):
        if (abs(pwInc) <= self.__maxPulse - self.__minPulse) and (pwInc != 0):
            self.__pwInc = pwInc
            return True
        else:
            print("Pulse Increment cannot be set to " + pwInc + ".")
            return False

    def increment(self):
        """'increment' function adds the value of pwInc to pulse"""
        if(self.__pulse + self.__pwInc <= self.__maxPulse):
            self.__pulse += self.__pwInc
            return True
        else:
            return False

    def decrement(self):
        """'decrement' function subtracts the value of pwInc from pulse"""
        if (self.__pulse - self.__pwInc >= self.__maxPulse):
            self.__pulse -= self.__pwInc
            return True
        else:
            return False

    def servoLeft(self):
        """'servoLeft' function increases the value of pulse by pwInc, effectively turning the servo to the left.  This does not pulse the servo.  A call to 'pi.set_servo_pulsewidth()' must be made."""
        if(self.__pwInc < 0):
            self.__pulse += (self.__pwInc * -1)
        else:
            self.__pulse += self.__pwInc

    def servoRight(self):
        """'servoRight' function decreases the value of pulse by pwInc, effectively turning the servo to the right.  This does not pulse the servo.  A call to 'pi.set_servo_pulsewidth()' must be made."""
        if (self.__pwInc > 0):
            self.__pulse += (self.__pwInc * -1)
        else:
            self.__pulse += self.__pwInc

    def __eq__(self, other):
        if(self.__gpio == other.__gpio) and (self.__pulse == other.__pulse) and (self.__maxPulse == other.__maxPulse) and (self.__minPulse == other.__minPulse) and (self.__pwInc == other.__pwInc):
            return True
        else:
            return False

    def __ne__(self, other):
        if (not self.__eq__(other)):
            return True
        else:
            return False


